### ISSUE TITLE
how to get correct outputs for each layer about 8-bit inception_v4 model

### ISSUE THREAD

**Issue by jetsmith** [opencv/dldt#147]
i am trying to output  the result of intermediate layers of 8-bit inception_v4 model which is produced by calibration tool. i used the project --- dldt/inference-engine/samples/classification_sample , the following code was added:
`std::vector<string> layer_name;
for (auto &&layer : network) {
std::string layerType = network.getLayerByName(layer->name.c_str())->type;
 {
 if (/*layerType != "Split" &&*/layerType != "Input" && layerType != "Const") {
network.addOutput(layer->name);
 layer_name.push_back(layer->name);
}
}
`
Then, outputs of intermediate layers could be got by code `const Blob::Ptr output_blob = infer_request.GetBlob();`.
However, the result after mixed_7d layer(include) are normal and the outputs before the layer are all anomalous, anomalous results like this:
(InceptionV4_InceptionV4_Conv2d_1a_3x3_Relu layer)
2.29596e-38
6.8395e-33
-2.76198e+36
5.61629e-30
7.25307e-30
2.15924e-29
-nan
-2.5246e+33
6.83705e-33
-4.92739e+17
9.21899e-38
-1.62888e+35
-2.18244e+24
1.51597e-36
2.56447e-23
3.70299e-37
-3.85297e+28
-3.88174e+31
6.23793e+32

i also test the float32 model which could get normal results, and i am sure my 8-bit inception_v4 model was produced correctly(79.85% top1 acc on imagenet). so how can i solve it?

**Comment by shubha-ramani** [comment]
Dear @jetsmith 
May I ask, how did you get your "8-bit model" ? We definitely support INT8 precision. But in order to use it you must convert your model to INT8 using the calibration tool. Before using the calibration tool, however,  you must generate a *.json and a *.pickle using convert_annotation.py, where you have to know which dataset you want to use. Please read here for additional information on annotation converters:
[https://github.com/opencv/dldt/blob/2019/tools/accuracy_checker/accuracy_checker/annotation_converters/README.md](https://github.com/opencv/dldt/blob/2019/tools/accuracy_checker/accuracy_checker/annotation_converters/README.md)

So the steps to do that are
1) [https://github.com/opencv/dldt/blob/2019/inference-engine/tools/accuracy_checker_tool/convert_annotation.py](https://github.com/opencv/dldt/blob/2019/inference-engine/tools/accuracy_checker_tool/convert_annotation.py)

2)[https://github.com/opencv/dldt/blob/2019/inference-engine/tools/calibration_tool/calibrate.py](https://github.com/opencv/dldt/blob/2019/inference-engine/tools/calibration_tool/calibrate.py)

3)[https://github.com/opencv/dldt/blob/2019/inference-engine/tools/accuracy_checker_tool/accuracy_check.py](https://github.com/opencv/dldt/blob/2019/inference-engine/tools/accuracy_checker_tool/accuracy_check.py)

So in summary, step 1) will create a *.json and a *.pickle file which will be consumed by Step 2) in the form of a "definitions.yml" file. Then in Step 3) you check the accuracy. Inception v4 is definitely supported.

You can get some sample configuration (*.yml) files here:
[https://software.intel.com/en-us/forums/computer-vision/topic/807243](https://software.intel.com/en-us/forums/computer-vision/topic/807243)

I know it's a lot of information. But the basic idea is that until you use our tools to generate proper INT8 IR, INT8 precision cannot be supported in OpenVino.

Hope this helps, and please let me know if you have other questions.

Thanks,

Shubha

**Comment by jetsmith** [comment]
Dear @shubha-ramani 
thanks for your detailed instructions, here are my calibration steps:
**1. Converting a TensorFlow Model**
under the dldt/model-optimizer, run mo_tf.py, here is the running shell:
`python3 mo_tf.py \
--input_model ./inception_v4.pb \
--model_name Inception_v4 \
--output_dir inception_v4 \
--input_shape [1,299,299,3] \
--data_type FP32 \
--mean_value [127.5,127.5,127.5] \
--scale 127.5 \
--reverse_input_channels \
--input input_img \
--output InceptionV4/Logits/Logits/BiasAdd
`

this step generates three files:
Inception_v4.bin  Inception_v4.mapping  Inception_v4.xml

2. **Calibration**
first build the project under dldt/inference-engine/, then calibration binaries could be found in inference-engine/bin/intel64/Release, then run the shell:
`./calibration_tool -t C \
	-i *****/dataset/imagenet_val/5000_image/img_class.txt \
	-m *****/dldt/model-optimizer/inception_v4_tf/Inception_v4.xml  \
	-d CPU \
	-subset 2000
`
Well, it produce **8-bit model** including Inception_v4_i8.bin and Inception_v4_i8.xml.
here is the calibration log:

[ INFO ] InferenceEngine: 
	API version ............ 1.6
	Build .................. custom_2019_b235c73481a6f133e1d182d91e5e2c29e8a460e5
[ INFO ] Parsing input parameters
[ INFO ] Loading plugin

	API version ............ 1.6
	Build .................. custom_2019_b235c73481a6f133e1d182d91e5e2c29e8a460e5
	Description ....... MKLDNNPlugin
[ INFO ] Loading network files
[ INFO ] Preparing input blobs
[ INFO ] Batch size is 1
[ INFO ] Collecting accuracy metric in FP32 mode to get a baseline, collecting activation statistics
...
FP32 accuracy: 79.95% vs current Int8 configuration accuracy: 79.85% with threshold for activation statistic: 97.50%
Layers profile for Int8 quantization
...
Write calibrated network to ****/dldt/model-optimizer/inception_v4_tf/Inception_v4_i8.(xml|bin) IR file

**Comment by jetsmith** [comment]
@shubha-ramani there is no inception_v4.yml in your sample configuration (*.yml) files link[https://software.intel.com/en-us/forums/computer-vision/topic/807243](url)

**Comment by shubha-ramani** [comment]
Here you go @jetsmith:

# http://nncv-jenkins.inn.intel.com/job/model_optimizer/job/private/job/ematroso_perf/job/run-accuracy-checker/26/artifact/tests/accuracy/reports/tf/INT8/inception_v4.log

# accuracy_check --progress print -d ./tests/definitions_full.yml -c ./tests/tf/INT8/inception_v4.yml -m /nfs/inn/proj/vdp/vdp_tests/models/internal/tf/1.12.0 --tf_custom_op_config_dir ./tf_custom_op_configs -M /home/jenkins/workspace/model_optimizer/private/ematroso_perf/run-accuracy-checker/model-optimizer -s /nfs/inn/proj/vdp/vdp_tests/datasets -a /nfs/inn/proj/vdp/vdp_tests/datasets -C /home/jenkins/workspace/model_optimizer/private/ematroso_perf/run-accuracy-checker/out_dir -e /usr/lib/x86_64-linux-gnu/ -tf dlsdk -td CPU -am ./tests/tf

models:
  - name: GoogleNet_v4

    launchers:
      - framework: caffe
        device: CPU
        caffe_model:   inception_v4/inception_v4.prototxt
        caffe_weights: inception_v4/inception_v4.caffemodel
        adapter: classification

      - framework: dlsdk
        device: CPU
        caffe_model:   inception_v4/inception_v4.prototxt
        caffe_weights: inception_v4/inception_v4.caffemodel
        adapter: classification
        mo_params:
            data_type: FP32

      - framework: dlsdk
        device: GPU
        caffe_model:   inception_v4/inception_v4.prototxt
        caffe_weights: inception_v4/inception_v4.caffemodel
        adapter: classification
        mo_params:
            data_type: FP32

      - framework: dlsdk
        device: HETERO:FPGA,CPU
        caffe_model:   inception_v4/inception_v4.prototxt
        caffe_weights: inception_v4/inception_v4.caffemodel
        adapter: classification
        mo_params:
            data_type: FP32

    datasets:
      - name: classification_dataset

        preprocessing:
        - type: resize
          size: 320
        - type: crop
          size: 299
        - type: normalization
          std: 255

**Comment by jetsmith** [comment]
@shubha-ramani many thanks.
